internal/state/tx_filter.go

```diff
@@ -1,22 +1,85 @@
 package state
 
 import (
+	"sync"
+	"time"
+
+	abci "github.com/tendermint/tendermint/abci/types"
 	"github.com/tendermint/tendermint/internal/mempool"
 	"github.com/tendermint/tendermint/types"
 )
 
-// TxPreCheck returns a function to filter transactions before processing.
-// The function limits the size of a transaction to the block's maximum data size.
-func TxPreCheck(state State) mempool.PreCheckFunc {
-	maxDataBytes := types.MaxDataBytesNoEvidence(
-		state.ConsensusParams.Block.MaxBytes,
-		state.Validators.Size(),
+func cachingStateFetcher(store Store) func() (State, error) {
```

Comment:
This is an improvement since it looks like we didn't update the check in the mempool at all in the previous code. I do think a height-based approach would be a bit better. We could update the `PostCheckTx` and `PreCheckTx` methods to take the height of the chain and have the caching fetcher update when it sees a new height.

Reply:
the Update method more or less does this, though the state store object doesn't allow you to access previous states so the actual final design here is something we should think more about. 

I think in general we should pass a store around in places where we currently pass a state (particularly during construction/startup, to avoid needing to connect to the app until we start (which is the core of #6777), so that's what this change does.

Having said that, the cache is probably a bit over wrought, but I think calling Load() for every single tx is probably also likely to cause a problem. And the case where Update isn't called on the mempool wouldn't behave meaingfully worse than the current implementation.